packages/components/src/tree-grid/index.js

```diff
@@ -0,0 +1,157 @@
+		const focusablesInRow = getRowFocusables( activeRow );
+		const currentColumnIndex = focusablesInRow.indexOf( activeElement );
+
+		if ( includes( [ LEFT, RIGHT ], keyCode ) ) {
+			// Calculate to the next element.
+			let nextIndex;
+			if ( keyCode === LEFT ) {
+				nextIndex = Math.max( 0, currentColumnIndex - 1 );
+			} else {
+				nextIndex = Math.min(
+					currentColumnIndex + 1,
+					focusablesInRow.length - 1
+				);
+			}
+
+			// Focus is either at the left or right edge of the grid. Do nothing.
+			if ( nextIndex === currentColumnIndex ) {
+				// Prevent key use for anything else. For example, Voiceover
+				// will start reading text on continued use of left/right arrow
+				// keys.
+				event.preventDefault();
+				return;
+			}
+
+			// Focus the next element.
+			focusablesInRow[ nextIndex ].focus();
+
+			// Prevent key use for anything else. This ensures Voiceover
+			// doesn't try to handle key navigation.
+			event.preventDefault();
+		} else if ( includes( [ UP, DOWN ], keyCode ) ) {
+			// Calculate the rowIndex of the next row.
+			const rows = Array.from(
+				treeGridElement.querySelectorAll( '[role="row"]' )
+			);
+			const currentRowIndex = rows.indexOf( activeRow );
+			let nextRowIndex;
+
+			if ( keyCode === UP ) {
+				nextRowIndex = Math.max( 0, currentRowIndex - 1 );
+			} else {
+				nextRowIndex = Math.min( currentRowIndex + 1, rows.length - 1 );
+			}
+
+			// Focus is either at the top or bottom edge of the grid. Do nothing.
+			if ( nextRowIndex === currentRowIndex ) {
+				// Prevent key use for anything else. For example, Voiceover
+				// will start navigating horizontally when reaching the vertical
+				// bounds of a table.
+				event.preventDefault();
+				return;
+			}
+
+			// Get the focusables in the next row.
+			const focusablesInNextRow = getRowFocusables(
+				rows[ nextRowIndex ]
+			);
+
+			// If for some reason there are no focusables in the next row, do nothing.
+			if ( ! focusablesInNextRow || ! focusablesInNextRow.length ) {
+				// Prevent key use for anything else. For example, Voiceover
+				// will still focus text when using arrow keys, while this
+				// component should limit navigation to focusables.
+				event.preventDefault();
+				return;
+			}
+
+			// Try to focus the element in the next row that's at a similar column to the activeElement.
+			const nextIndex = Math.min(
+				currentColumnIndex,
+				focusablesInNextRow.length - 1
+			);
+			focusablesInNextRow[ nextIndex ].focus();
+
+			// Prevent key use for anything else. This ensures Voiceover
+			// doesn't try to handle key navigation.
+			event.preventDefault();
+		}
+	}, [] );
+
+	return (
+		<RovingTabIndexContainer>
+			{ /* Disable reason: A treegrid is implemented using a table element. */ }
+			{ /* eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role */ }
+			<table { ...props } role="treegrid" onKeyDown={ onKeyDown }>
```

Comment:
Might be good to add in an `aria-label` here that can be passed a short description of what the treegrid is about.